File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return False

        visited = set()

        cur = head
        visited.add(cur)
        while cur.next is not None:
            if cur.next in visited:
                return True

            cur = cur.next
            visited.add(cur)

        return False

File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {ListNode}
     */
    reverseList(head) {
        if (head == null) return null

        /**
         * @param {ListNode} prev
         * @param {ListNode} cur
         */
        const reverse = (prev, cur, i) => {
            if (cur == null) {
                const prev2 = new ListNode(prev.val)
                return [prev2, prev2]
            }
            const [tail, cur2] = reverse(cur, cur.next, i + 1)
            const prev2 = new ListNode(prev.val)
            cur2.next = prev2
            return [tail, prev2]
        }
        const tail = reverse(head, head.next, 0)[0]
        return tail
    }
}
